src/ecephys_etl/modules/vbn_create_stimulus_table/create_stim_table.py

```diff
@@ -0,0 +1,551 @@
+            f"is not guaranteed to be correct!"
+        )
+
+    # 'stim_running'/'sweep' line is high while visual stimulus is being
+    # displayed and low otherwise
+    stim_starts = sync_dataset.get_rising_edges(stim_line, units='seconds')
+    stim_ends = sync_dataset.get_falling_edges(stim_line, units='seconds')
+
+    return stim_starts, stim_ends
+
+
+def get_frame_offsets(
+    sync_dataset: Dataset, frame_counts: List[int], tolerance: int = 0
+) -> List[int]:
+    """List of the inferred start frames for each of the stimuli
+
+    Parameters
+    ----------
+    sync_dataset : Dataset
+        Sync data from an ecephys session (a 'Dataset' object that has loaded
+        a *.sync H5 file).
+    frame_counts : List[int]
+        List of expected frame counts (taken from behavior, mapping, and replay
+        pkl files) for each stimuli category. The list of counts should be
+        ordered by the actual display sequence of the stimuli categories.
+    tolerance : int
+        Percent by which frame counts are allowed to deviate from,
+        by default 0.
+
+    Returns
+    -------
+    List[float]
+        List of the inferred start frames for each stimuli category presented
+        during session.
+    """
+    frame_count_arr = np.array(frame_counts)
+    tolerance_pct = tolerance / 100.
+
+    # Get vsyncs and stim_running signals from sync data
+    vsync_times = get_vsyncs(sync_dataset)
+    stim_starts, stim_ends = get_stim_starts_and_ends(sync_dataset)
+
+    # Get vsync frame lengths and indices for all stimuli
+    epoch_frame_counts = []
+    epoch_start_frames = []
+    for start, end in zip(stim_starts, stim_ends):
+        # Inner expression returns a bool array where conditions are True
+        # np.where evaluates bool array to return indices where bool array True
+        epoch_frames = np.where((vsync_times > start) & (vsync_times < end))[0]
+        epoch_frame_counts.append(len(epoch_frames))
+        epoch_start_frames.append(epoch_frames[0])
+
+    if len(epoch_frame_counts) == len(frame_count_arr):
+        if not np.allclose(frame_count_arr, epoch_frame_counts):
+            logging.warning(
+                f"Number of frames derived from sync file "
+                f"({epoch_frame_counts})for each epoch not matching up with "
+                f"frame counts derived from pkl files ({frame_count_arr})!"
+            )
+        start_frames = epoch_start_frames
+    elif len(epoch_frame_counts) > len(frame_count_arr):
+        logging.warning(
+            f"Number of stim presentations obtained from sync "
+            f"({len(epoch_frame_counts)}) higher than number expected "
+            f"({len(frame_count_arr)}). Inferring start frames."
+        )
+
+        start_frames = []
+        for stim_idx, fc in enumerate(frame_count_arr):
+
+            logging.info(f"Finding stim start for stim with index: {stim_idx}")
+            # Get index of stimulus whose frame counts most closely match
```

Comment:
is this robust? doesn't this rely on stimuli having differing numbers of frame counts?

Reply:
This is robust if `frame_count_arr` is in order that was actually presented. I will add a warning note to the docstring for this function.